source/libuv/uv_event_loop.c

```diff
@@ -0,0 +1,792 @@
+    }
+
+    if (s_is_on_callers_thread(event_loop)) {
+        /* If on UV thread, directly start sub */
+        uv_poll_init(impl->uv_loop, &handle_data->poll, handle_data->owner->data.fd);
+        uv_poll_start(&handle_data->poll, handle_data->uv_events, s_uv_poll_cb);
+    } else {
+        /* Otherwise, schedule async to do sub */
+        aws_atomic_fetch_add(&impl->num_open_handles, 1);
+
+        handle_data->poll_start_async.data = handle_data;
+        uv_async_init(impl->uv_loop, &handle_data->poll_start_async, s_uv_async_poll_start);
+        uv_async_send(&handle_data->poll_start_async);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_unsubscribe_from_io_events(struct aws_event_loop *event_loop, struct aws_io_handle *handle) {
+
+    (void)event_loop;
+    assert(handle->additional_data);
+
+    struct handle_data *handle_data = handle->additional_data;
+
+    if (uv_poll_stop(&handle_data->poll)) {
+        return AWS_OP_ERR;
+    }
+    uv_close((uv_handle_t *)&handle_data->poll, s_uv_close_sub);
+
+    handle->additional_data = NULL;
+
+    aws_linked_list_remove(&handle_data->open_subs_node);
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_free_io_event_resources(void *user_data) {
+
+    struct handle_data *handle_data = user_data;
+
+    aws_mem_release(handle_data->event_loop->alloc, handle_data);
+}
+
+static bool s_is_on_callers_thread(struct aws_event_loop *event_loop) {
+    struct libuv_loop *impl = event_loop->impl_data;
+
+    const uint64_t uv_tid =
+        impl->owns_uv_loop ? aws_thread_get_id(&s_owned(impl)->thread) : s_unowned(impl)->uv_thread_id;
+
+    return uv_tid == aws_thread_current_thread_id();
+}
+
+struct aws_event_loop_vtable s_libuv_vtable = {
+    .destroy = s_destroy,
+    .run = s_run,
+    .stop = s_stop,
+    .wait_for_stop_completion = s_wait_for_stop_completion,
+    .schedule_task_now = s_schedule_task_now,
+    .schedule_task_future = s_schedule_task_future,
+    .cancel_task = s_cancel_task,
+    .subscribe_to_io_events = s_subscribe_to_io_events,
+    .unsubscribe_from_io_events = s_unsubscribe_from_io_events,
+    .free_io_event_resources = s_free_io_event_resources,
+    .is_on_callers_thread = s_is_on_callers_thread,
+};
+
+static int s_libuv_loop_init(
+    struct aws_event_loop *event_loop,
+    struct aws_allocator *alloc,
+    struct libuv_loop *impl,
+    uv_loop_t *uv_loop,
+    aws_io_clock_fn *clock) {
+
+    if (aws_event_loop_init_base(event_loop, alloc, clock)) {
+        return AWS_OP_ERR;
+    }
+
+    impl->uv_loop = uv_loop;
+    aws_atomic_init_int(&impl->num_open_handles, 0);
+
+    /* Init cross-thread data */
+    if (aws_mutex_init(&impl->cross_thread_data.mutex)) {
+        goto clean_up;
+    }
+
+    aws_linked_list_init(&impl->cross_thread_data.tasks_to_schedule);
+
+    /* Init on-thread data */
+    if (aws_hash_table_init(&impl->el_thread_data.running_tasks, alloc, 5, aws_hash_ptr, aws_ptr_eq, NULL, NULL)) {
+        goto clean_up;
```

Comment:
clean_up isn't cleaning up hash table

Reply:
Correct, there is no `goto clean_up` in a case where this hash table was successfully created.